### PR TITLE
Add support for apps

### PR DESCRIPTION
**Pull request by rxl881**

### CODE REVIEW

File: src/UserSettingsStore.js [hunk 1]
```diff
@@ -30,6 +30,11 @@ export default {
             id: 'rich_text_editor',
             default: false,
         },
+        {
+            name: "Matrix Apps",
```
Comment (Kegsay) [inline src/UserSettingsStore.js]: Please add a translation as per `doTranslations` below.

File: src/components/structures/RoomView.js [hunk 2]
```diff
@@ -47,13 +47,12 @@ import UserProvider from '../../autocomplete/UserProvider';
 
 import RoomViewStore from '../../stores/RoomViewStore';
 
-var DEBUG = false;
+const DEBUG = false;
```
Comment (Kegsay) [inline src/components/structures/RoomView.js]: I don't like this. If we attach a debugger we can't turn on `DEBUG` at run-time with this change.

File: src/components/structures/RoomView.js [hunk 3]
```diff
@@ -91,6 +90,7 @@ module.exports = React.createClass({
         return {
             room: null,
             roomId: null,
+            userId: null,
```
Comment (Kegsay) [inline src/components/structures/RoomView.js]: Looks like we don't use this anywhere? We yank it from `MatrixClientPeg.get().credentials.userId` on `:1619`.

File: src/components/structures/RoomView.js [hunk 4]
```diff
@@ -453,10 +460,15 @@ module.exports = React.createClass({
...
 
                 this.setState({
-                    callState: callState
+                    callState: callState,
                 });
 
                 break;
+            case 'appsDrawer':
+                this.setState({
+                    showApps: payload.show ? true : false,
```
Comment (Kegsay) [inline src/components/structures/RoomView.js]: AKA:
```js
showApps: payload.show
```

File: src/components/views/elements/AppTile.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
+/*
+Copyright 2015, 2016 OpenMarket Ltd
```
Comment (Kegsay) [inline src/components/views/elements/AppTile.js]: 2017 Vector Creations please.

File: src/components/views/elements/AppTile.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+                            src="img/edit.svg"
+                            className="mx_filterFlipColor mx_AppTileMenuBarWidget mx_AppTileMenuBarWidgetPadding"
+                            width="8" height="8" alt="Edit"
+                            onClick={this._onEditClick}
+                        /> */}
+
+                        {/* Delete widget */}
+                        <img src="img/cancel.svg"
+                        className="mx_filterFlipColor mx_AppTileMenuBarWidget"
+                        width="8" height="8" alt="Cancel"
```
Comment (Kegsay) [inline src/components/views/elements/AppTile.js]: i18n `"Cancel"` please.

File: src/components/views/rooms/AppsDrawer.js [hunk 7]
```diff
@@ -0,0 +1,220 @@
+/*
+Copyright 2015, 2016 OpenMarket Ltd
```
Comment (Kegsay) [inline src/components/views/rooms/AppsDrawer.js]: 2017 Vector Creations

File: src/components/views/rooms/AppsDrawer.js [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+    },
+
+    componentWillMount: function() {
+        ScalarMessaging.startListening();
+        MatrixClientPeg.get().on("RoomState.events", this.onRoomStateEvents);
+    },
+
+    componentDidMount: function() {
+        this.scalarClient = null;
+        const appsDrawer = this;
```
Comment (Kegsay) [inline src/components/views/rooms/AppsDrawer.js]: Please don't do this, we have fat-arrow functions, let's just use `this` throughout.

File: src/components/views/rooms/AppsDrawer.js [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+        //         app.queryParams = '?roomAlias=' + encodeURIComponent(app.data.roomAlias);
+        //         break;
+        // }
+
+        return app;
+    },
+
+    getInitialState: function() {
+        return {
+            apps: this._getApps(),
```
Comment (Kegsay) [inline src/components/views/rooms/AppsDrawer.js]: If you can put this right near the top along with `propTypes` that'd be great so I can see what variables will exist in `state`.

File: src/components/views/rooms/AppsDrawer.js [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+        if (SdkConfig.get().integrations_ui_url && SdkConfig.get().integrations_rest_url) {
+            this.scalarClient = new ScalarAuthClient();
+            this.scalarClient.connect().done(() => {
+                this.forceUpdate();
+                if (appsDrawer.state.apps && appsDrawer.state.apps.length < 1) {
+                    appsDrawer.onClickAddWidget();
+                }
+            }, (err) => {
+                this.setState({
+                    scalar_error: err,
```
Comment (Kegsay) [inline src/components/views/rooms/AppsDrawer.js]: You never read this.

File: src/components/views/rooms/AppsDrawer.js [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+    _updateApps: function() {
+        const apps = this._getApps();
+        if (apps.length < 1) {
+            dis.dispatch({
+                action: 'appsDrawer',
+                show: false,
+            });
+        }
+        this.setState({
+            apps: this._getApps(),
```
Comment (Kegsay) [inline src/components/views/rooms/AppsDrawer.js]: Why do we call `_getApps()` twice? Can we just reference `apps`?

File: src/components/views/rooms/AppsDrawer.js [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+                />;
+            });
+
+        const addWidget = this.state.apps && this.state.apps.length < 2 &&
+            (<div onClick={this.onClickAddWidget}
+                            role="button"
+                            tabIndex="0"
+                            className="mx_AddWidget_button"
+                            title="Add a widget">
+                            [+] Add a widget
```
Comment (Kegsay) [inline src/components/views/rooms/AppsDrawer.js]: i18n this please.

File: src/components/views/rooms/MessageComposer.js [hunk 13]
```diff
@@ -267,14 +314,29 @@ export default class MessageComposer extends React.Component {
                 </div>;
         }
 
-        var canSendMessages = this.props.room.currentState.maySendMessage(
+        // Apps
+        if (UserSettingsStore.isFeatureEnabled('matrix_apps')) {
+            if (this.props.showApps) {
+                hideAppsButton =
+                    <div key="controls_hide_apps" className="mx_MessageComposer_apps" onClick={this.onHideAppsClick} title="Hide Apps">
```
Comment (Kegsay) [inline src/components/views/rooms/MessageComposer.js]: i18n "Hide Apps" please.

File: src/components/views/rooms/MessageComposer.js [hunk 14]
```diff
@@ -267,14 +314,29 @@ export default class MessageComposer extends React.Component {
...
+                        <TintableSvg src="img/icons-apps-active.svg" width="35" height="35"/>
+                    </div>;
+            } else {
+                showAppsButton =
+                    <div key="show_apps" className="mx_MessageComposer_apps" onClick={this.onShowAppsClick} title="Show Apps">
```
Comment (Kegsay) [inline src/components/views/rooms/MessageComposer.js]: i18n "Show Apps" please.

**Comment by Kegsay** [comment]
Otherwise LGTM.

File: src/UserSettingsStore.js (hunk 1 above)
Comment (rxl881) [inline src/UserSettingsStore.js]: Thanks. Done.

File: src/components/structures/RoomView.js (hunk 2 above)
Comment (rxl881) [inline src/components/structures/RoomView.js]: Cheers, I hadn't considered that case. I've changed it to 'let'.

File: src/components/structures/RoomView.js (hunk 3 above)
Comment (rxl881) [inline src/components/structures/RoomView.js]: Hmm... interesting ... it looks like the state was being used in my commit - https://github.com/matrix-org/matrix-react-sdk/commit/dc4f321707da8d7431c857ab6abf8240e671fb31. I'm not sure how this got changed, but I've removed it from the state in preference for accessing the value directly, as you pointed out in :1619

File: src/components/structures/RoomView.js (hunk 4 above)
Comment (rxl881) [inline src/components/structures/RoomView.js]: Ahh, yeah, good point! :)

File: src/components/views/elements/AppTile.js [hunk 15]
```diff
@@ -87,7 +87,7 @@ export default React.createClass({
                         {/* Delete widget */}
                         <img src="img/cancel.svg"
                         className="mx_filterFlipColor mx_AppTileMenuBarWidget"
-                        width="8" height="8" alt="Cancel"
+                        width="8" height="8" alt={_t("Cancel")}
```
Comment (Kegsay) [inline src/components/views/elements/AppTile.js]: You need to import `_t`.

**Comment by rxl881** [comment]
Thanks for the review @Kegsay. All fixes should be applied now.

File: src/components/views/rooms/AppsDrawer.js [hunk 16]
```diff
@@ -163,7 +163,7 @@ module.exports = React.createClass({
             });
         }
         this.setState({
-            apps: this._getApps(),
+            apps: this.apps,
```
Comment (Kegsay) [inline src/components/views/rooms/AppsDrawer.js]: This should be `apps` from `:158`.

**Comment by Kegsay** [comment]
- Missing `_t` import in `src/components/views/elements/AppTile.js`
- `apps: this.apps,` should be `apps: apps,` in `src/components/views/rooms/AppsDrawer.js`

Otherwise LGTM

File: src/components/views/rooms/AppsDrawer.js (hunk 16 above)
Comment (rxl881) [inline src/components/views/rooms/AppsDrawer.js]: Great spot!

File: src/components/views/rooms/AppsDrawer.js (hunk 9 above)
Comment (rxl881) [inline src/components/views/rooms/AppsDrawer.js]: Done

**Comment by Kegsay** [comment]
LGTM

File: src/components/views/rooms/AppsDrawer.js [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+        return pathTemplate;
+    },
+
+    _initAppConfig: function(appId, app) {
+        const user = MatrixClientPeg.get().getUser(this.props.userId);
+        const params = {
+            '$matrix_user_id': this.props.userId,
+            '$matrix_room_id': this.props.room.roomId,
+            '$matrix_display_name': user ? user.displayName : this.props.userId,
+            '$matrix_avatar_url': user ? MatrixClientPeg.get().mxcUrlToHttp(user.avatarUrl) : '',
```
Comment (dbkr) [inline src/components/views/rooms/AppsDrawer.js]: I don't think you should be de-mxcing here: you'll be effectively directing the scalar app to download your avatar from the user's own HS, and this won't always be possible.

**Pull request merged**